File: packages/aegisops-python/src/aegisops/middleware.py
from __future__ import annotations

import time
import uuid
from typing import Any, Dict, Iterable, Optional, Tuple

from .client import AegisOpsClient, AegisOpsConfig, create_aegisops_client


HeaderList = Iterable[Tuple[bytes, bytes]]


class AegisOpsMiddleware:
    def __init__(
        self,
        app,
        config: Optional[AegisOpsConfig] = None,
        client: Optional[AegisOpsClient] = None,
    ):
        self.app = app
        self.client = client or create_aegisops_client(config)

    async def __call__(self, scope: Dict[str, Any], receive, send):
        if scope.get("type") != "http":
            await self.app(scope, receive, send)
            return

        path = scope.get("path", "")
        if self.client.should_ignore_route(path):
            await self.app(scope, receive, send)
            return

        started = time.perf_counter()
        headers = _headers(scope.get("headers", []))
        request_id = headers.get("x-request-id") or headers.get("x-correlation-id") or str(uuid.uuid4())
        trace_id = headers.get("x-trace-id") or headers.get("traceparent") or str(uuid.uuid4())
        method = scope.get("method", "GET")
        route = path
        status_code = 500

        async def wrapped_send(message):
            nonlocal status_code
            if message.get("type") == "http.response.start":
                status_code = int(message.get("status", status_code))
                response_headers = list(message.get("headers", []))
                lower_names = {name.lower() for name, _value in response_headers}
                if b"x-request-id" not in lower_names:
                    response_headers.append((b"x-request-id", request_id.encode("utf-8")))
                if b"x-trace-id" not in lower_names:
                    response_headers.append((b"x-trace-id", trace_id.encode("utf-8")))
                message = {**message, "headers": response_headers}
            await send(message)

        try:
            await self.app(scope, receive, wrapped_send)
        except Exception as exc:
            duration_ms = _duration_ms(started)
            labels = _labels(method, route, 500)
            self.client.send_batch_metrics(
                [
                    {"metricName": "http_requests_total", "value": 1, "labels": labels},
                    {"metricName": "http_request_duration_ms", "value": duration_ms, "labels": labels},
                    {"metricName": "http_errors_total", "value": 1, "labels": labels},
                    {"metricName": "http_5xx_total", "value": 1, "labels": labels},
                    {"metricName": "exceptions_total", "value": 1, "labels": labels},
                ]
            )
            self.client.send_log(
                {
                    "level": "error",
                    "message": str(exc) or exc.__class__.__name__,
                    "requestId": request_id,
                    "traceId": trace_id,
                    "route": route,
                    "method": method,
                    "statusCode": 500,
                    "durationMs": duration_ms,
                    "metadata": _metadata(scope, route, 500, duration_ms, request_id, trace_id, self.client.config),
                }
            )
            raise

        duration_ms = _duration_ms(started)
        labels = _labels(method, route, status_code)
        metrics = [
            {"metricName": "http_requests_total", "value": 1, "labels": labels},
            {"metricName": "http_request_duration_ms", "value": duration_ms, "labels": labels},
        ]
        if status_code >= 400:
            metrics.append({"metricName": "http_errors_total", "value": 1, "labels": labels})
        if status_code >= 500:
            metrics.append({"metricName": "http_5xx_total", "value": 1, "labels": labels})
        elif status_code >= 400:
            metrics.append({"metricName": "http_4xx_total", "value": 1, "labels": labels})
        if duration_ms >= self.client.config.slow_request_threshold_ms:
            metrics.append({"metricName": "slow_requests_total", "value": 1, "labels": labels})
        self.client.send_batch_metrics(metrics)

        metadata = _metadata(scope, route, status_code, duration_ms, request_id, trace_id, self.client.config)
        if duration_ms >= self.client.config.slow_request_threshold_ms:
            self.client.send_log(
                {
                    "level": "warn",
                    "message": f"Slow request {method} {route}",
                    "requestId": request_id,
                    "traceId": trace_id,
                    "route": route,
                    "method": method,
                    "statusCode": status_code,
                    "durationMs": duration_ms,
                    "metadata": metadata,
                }
            )
        if status_code >= 500:
            self.client.send_log(
                {
                    "level": "error",
                    "message": f"HTTP {status_code} {method} {route}",
                    "requestId": request_id,
                    "traceId": trace_id,
                    "route": route,
                    "method": method,
                    "statusCode": status_code,
                    "durationMs": duration_ms,
                    "metadata": metadata,
                }
            )


def _headers(raw_headers: HeaderList) -> Dict[str, str]:
    return {
        name.decode("latin-1").lower(): value.decode("latin-1")
        for name, value in raw_headers
    }


def _labels(method: str, route: str, status_code: int) -> Dict[str, str]:
    return {"method": method, "route": route, "statusCode": str(status_code)}


def _duration_ms(started: float) -> float:
    return (time.perf_counter() - started) * 1000


def _metadata(
    scope: Dict[str, Any],
    route: str,
    status_code: int,
    duration_ms: float,
    request_id: str,
    trace_id: str,
    config: AegisOpsConfig,
) -> Dict[str, Any]:
    client = scope.get("client") or (None, None)
    headers = _headers(scope.get("headers", []))
    return {
        "route": route,
        "method": scope.get("method", "GET"),
        "statusCode": status_code,
        "durationMs": duration_ms,
        "requestId": request_id,
        "traceId": trace_id,
        "userAgent": headers.get("user-agent"),
        "ip": client[0],
        "projectKey": config.project_key,
        "serviceName": config.service_name,
        "environment": config.environment,
    }

File: packages/aegisops-go/middleware.go
package aegisops

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type responseRecorder struct {
	http.ResponseWriter
	statusCode  int
	wroteHeader bool
}

func (recorder *responseRecorder) WriteHeader(statusCode int) {
	if recorder.wroteHeader {
		return
	}
	recorder.statusCode = statusCode
	recorder.wroteHeader = true
	recorder.ResponseWriter.WriteHeader(statusCode)
}

func (recorder *responseRecorder) Write(data []byte) (int, error) {
	if !recorder.wroteHeader {
		recorder.WriteHeader(http.StatusOK)
	}
	return recorder.ResponseWriter.Write(data)
}

func Middleware(cfg Config) func(http.Handler) http.Handler {
	client := NewClient(cfg)
	return client.Middleware
}

func (client *Client) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path := r.URL.Path
		if client.shouldIgnoreRoute(path) {
			next.ServeHTTP(w, r)
			return
		}

		started := time.Now()
		requestID := firstNonEmpty(r.Header.Get("x-request-id"), r.Header.Get("x-correlation-id"), randomID())
		traceID := firstNonEmpty(r.Header.Get("x-trace-id"), r.Header.Get("traceparent"), randomID())
		w.Header().Set("x-request-id", requestID)
		w.Header().Set("x-trace-id", traceID)

		recorder := &responseRecorder{ResponseWriter: w, statusCode: http.StatusOK}
		defer func() {
			recovered := recover()
			if recovered != nil && !recorder.wroteHeader {
				http.Error(recorder, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
			}
			durationMS := float64(time.Since(started).Microseconds()) / 1000.0
			statusCode := recorder.statusCode
			if recovered != nil {
				statusCode = http.StatusInternalServerError
			}
			route := path
			labels := map[string]string{"method": r.Method, "route": route, "statusCode": intString(statusCode)}
			metrics := []Metric{
				{MetricName: "http_requests_total", Value: 1, Labels: labels},
				{MetricName: "http_request_duration_ms", Value: durationMS, Labels: labels},
			}
			if statusCode >= 400 {
				metrics = append(metrics, Metric{MetricName: "http_errors_total", Value: 1, Labels: labels})
			}
			if statusCode >= 500 {
				metrics = append(metrics, Metric{MetricName: "http_5xx_total", Value: 1, Labels: labels})
			} else if statusCode >= 400 {
				metrics = append(metrics, Metric{MetricName: "http_4xx_total", Value: 1, Labels: labels})
			}
			if time.Since(started) >= client.cfg.SlowRequestThreshold {
				metrics = append(metrics, Metric{MetricName: "slow_requests_total", Value: 1, Labels: labels})
			}
			if recovered != nil {
				metrics = append(metrics, Metric{MetricName: "exceptions_total", Value: 1, Labels: labels})
			}
			client.SendBatchMetrics(context.Background(), metrics)

			metadata := map[string]interface{}{
				"route":       route,
				"method":      r.Method,
				"statusCode":  statusCode,
				"durationMs":  durationMS,
				"requestId":   requestID,
				"traceId":     traceID,
				"userAgent":   r.UserAgent(),
				"ip":          clientIP(r),
				"projectKey":  client.cfg.ProjectKey,
				"serviceName": client.cfg.ServiceName,
				"environment": client.cfg.Environment,
			}
			if recovered != nil {
				metadata["panic"] = recovered
				client.SendLog(context.Background(), Log{
					Level:      "error",
					Message:    "Unhandled panic",
					RequestID:  requestID,
					TraceID:    traceID,
					Route:      route,
					Method:     r.Method,
					StatusCode: http.StatusInternalServerError,
					DurationMS: durationMS,
					Metadata:   metadata,
				})
				return
			}
			if time.Since(started) >= client.cfg.SlowRequestThreshold {
				client.SendLog(context.Background(), Log{
					Level:      "warn",
					Message:    "Slow request " + r.Method + " " + route,
					RequestID:  requestID,
					TraceID:    traceID,
					Route:      route,
					Method:     r.Method,
					StatusCode: statusCode,
					DurationMS: durationMS,
					Metadata:   metadata,
				})
			}
			if statusCode >= 500 {
				client.SendLog(context.Background(), Log{
					Level:      "error",
					Message:    "HTTP " + intString(statusCode) + " " + r.Method + " " + route,
					RequestID:  requestID,
					TraceID:    traceID,
					Route:      route,
					Method:     r.Method,
					StatusCode: statusCode,
					DurationMS: durationMS,
					Metadata:   metadata,
				})
			}
		}()

		next.ServeHTTP(recorder, r)
	})
}

func (client *Client) shouldIgnoreRoute(path string) bool {
	for _, route := range client.cfg.IgnoredRoutes {
		if path == route || strings.HasPrefix(path, strings.TrimRight(route, "/")+"/") {
			return true
		}
	}
	return false
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func randomID() string {
	var data [16]byte
	if _, err := rand.Read(data[:]); err != nil {
		return time.Now().UTC().Format("20060102150405.000000000")
	}
	return hex.EncodeToString(data[:])
}

func intString(value int) string {
	return strconv.Itoa(value)
}

func clientIP(r *http.Request) string {
	if forwarded := r.Header.Get("x-forwarded-for"); forwarded != "" {
		return strings.TrimSpace(strings.Split(forwarded, ",")[0])
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err == nil {
		return host
	}
	return r.RemoteAddr
}
